mkosi/__init__.py

```diff
@@ -362,7 +362,11 @@ def set_root_pw(line: str) -> str:
                     return ":".join(["root", password] + line.split(":")[2:])
                 return line
 
-            patch_file(state.root / "etc/shadow", set_root_pw)
+            shadow = state.root / "etc/shadow"
+            try:
+                patch_file(shadow, set_root_pw)
+            except FileNotFoundError:
+                shadow.write_text(f"root:{password}:0:0:99999:7:::")
```

Comment:
Bit late, but `Path.write_text` won't append a trailing newline, so the generated file will be missing that.